File: src/features/formazione/comuni.ts
import { MODULI_AZIENDA, type ModuloAzienda } from "@/features/companies/moduli";
import { NUMERI } from "./numeri";
import type { Sezione } from "./tipi";

/**
 * Le sezioni valide per TUTTI i percorsi, scritte una volta.
 *
 * ⚠️ Metà del contenuto di un corso è identica fra i dodici moduli: come ci si muove, come
 * si salva, che cosa controlla la verifica, che cosa succede quando si pubblica. Scritta
 * dodici volte diverge alla prima correzione, e a divergere è sempre la copia che nessuno
 * rilegge.
 *
 * ⚠️ E QUI NON SI NOMINA NIENTE CHE NON ESISTA. I due corsi da cui questo materiale
 * proviene descrivevano i PROTOTIPI: dicevano che i dati vivono nel browser, che il backup
 * è l'esportazione di un JSON, che cambiare computer significa perdere il lavoro, e che il
 * PDF si fa con la stampa del browser. In EvalisDeck nessuna di queste è vera, e la terza
 * è anche allarmante — insegnerebbe un rito inutile a chi ha i dati al sicuro su un
 * database. Un corso che sbaglia il prodotto è peggio di nessun corso: chi lo legge smette
 * di cercare altrove.
 */
export function sezioniComuni(modulo: ModuloAzienda): Sezione[] {
  const m = MODULI_AZIENDA.find((x) => x.href === modulo)!;
  const documento = m.documenti[0];

  return [
    {
      id: "dove-sei",
      titolo: "Dove sei, e come ci si muove",
      minuti: 3,
      sommario: "Tre livelli: il portafoglio, il fascicolo di un'azienda, il percorso che stai seguendo.",
      blocchi: [
        {
          tipo: "prosa",
          testo:
            `EvalisDeck si attraversa sempre nello stesso ordine. Il **portafoglio** elenca le aziende che segui. ` +
            `Aprendone una si entra nel suo **fascicolo**, che raccoglie i ${NUMERI.moduli} percorsi disponibili ` +
            `divisi in ${NUMERI.gruppi} gruppi. Da lì si apre il percorso su cui si vuole lavorare.`,
        },
        {
          tipo: "prosa",
          testo:
            `La barra laterale segue questa struttura: quando sei dentro un'azienda mostra il suo nome e i suoi ` +
            `percorsi, così passi dal Bilancio alla SoA della stessa azienda senza tornare indietro.`,
        },
        {
          tipo: "avviso",
          tono: "nota",
          titolo: "I percorsi che non apri non ti disturbano",
          testo:
            `Un'azienda ha ${NUMERI.moduli} percorsi disponibili, non ${NUMERI.moduli} percorsi da fare. Si aprono ` +
            `quelli che servono a quel cliente: gli altri restano nel fascicolo, spenti, e non compaiono nello ` +
            `scadenzario.`,
        },
      ],
    },

    {
      id: "come-si-salva",
      titolo: "Come si salva",
      minuti: 3,
      sommario: "Da solo, campo per campo. Non c'è un pulsante «salva», e non c'è niente da esportare.",
      blocchi: [
        {
          tipo: "prosa",
          testo:
            `Ogni campo si salva quando lo lasci: compare per un istante la conferma accanto al campo, e il dato ` +
            `è sul server. Non esiste un pulsante «salva», e non esiste un modo di perdere il lavoro chiudendo la ` +
            `pagina.`,
        },
        {
          tipo: "avviso",
          tono: "attenzione",
          titolo: "Il salvataggio è per CAMPO, e conta",
          testo:
            `Ogni modifica riguarda il solo campo che hai toccato, mai la riga intera. È una scelta pagata con ` +
            `quattro difetti veri: salvare il costo azzerava la quantità, impostare la rilevanza finanziaria ` +
            `cancellava l'impatto appena messo. Per te significa che puoi compilare in fretta, anche saltando da ` +
            `un campo all'altro, senza che l'ultimo sovrascriva il penultimo.`,
        },
        {
          tipo: "prosa",
          testo:
            `I dati di ogni studio sono isolati **a livello di database**: un altro studio non può leggerli nemmeno ` +
            `per errore di programmazione. I colleghi che inviti nel tuo studio vedono invece tutto il portafoglio, ` +
            `perché è lavoro condiviso.`,
        },
        {
          tipo: "avviso",
          tono: "nota",
          titolo: "Niente esportazioni di sicurezza",
          testo:
            `Non c'è nessun archivio da esportare e nessun file da conservare: il lavoro sta sul server, e cambiare ` +
            `computer significa solo rifare l'accesso. Se hai usato i prototipi, è il cambiamento più grande: lì il ` +
            `backup era compito tuo.`,
        },
      ],
    },

    esercizio(m.perEsercizio),

    {
      id: "la-verifica",
      titolo: "La verifica prima di consegnare",
      minuti: 3,
      sommario: "L'elenco di cosa manca, con il collegamento che porta al punto da sistemare.",
      blocchi: [
        {
          tipo: "prosa",
          testo:
            `Il passo di verifica non è un riassunto: è la lista di ciò che impedisce di consegnare. Ogni voce ` +
            `porta al punto esatto da correggere, così non devi cercarlo.`,
        },
        {
          tipo: "avviso",
          tono: "errore",
          titolo: "Verde non vuol dire corretto",
          testo:
            `La verifica controlla la **completezza**, non la **correttezza**: un numero sbagliato ma presente ` +
            `risulta a posto. La quadratura con le fatture, la plausibilità dei fattori di carico, la coerenza dei ` +
            `delta da un anno all'altro restano lavoro tuo. Nessun software può firmare al posto di chi firma.`,
        },
      ],
    },

    {
      id: "pubblicare",
      titolo: "Che cosa succede quando pubblichi",
      minuti: 4,
      sommario: "Il documento si congela. Da quel momento non può più cambiare, ed è il motivo per cui vale.",
      blocchi: [
        {
          tipo: "prosa",
          testo:
            `Pubblicare non significa «stampare». Significa congelare in una versione numerata i dati **e i calcoli** ` +
            `che li accompagnano: da quel momento il documento non cambia più, nemmeno se domani correggi un dato ` +
            `del percorso. Il divieto è imposto dal database, non è una promessa dell'applicazione.`,
        },
        {
          tipo: "elenco",
          voci: [
            "Il documento prende un **numero di versione**: ripubblicando ottieni la v2, e la v1 resta consultabile.",
            "Il **marchio** si sceglie in quel momento e resta quello: i documenti già consegnati non cambiano intestazione.",
            "L'**edizione dei contenuti** su cui è stato redatto viene congelata: le norme si aggiornano, e chi riceve il documento deve poter sapere su quale versione è stato costruito.",
            "Il documento riceve un **codice di verifica pubblico**: chi lo ha in mano può controllarne l'autenticità da una pagina del sito, senza account.",
          ],
        },
        {
          tipo: "prosa",
          testo:
            `Il PDF si genera una volta e resta archiviato: riscaricarlo restituisce lo stesso file, byte per byte. ` +
            `È coerente con l'immutabilità — quello consegnato al cliente è quello, e non può cambiare impaginazione ` +
            `sotto i suoi occhi.`,
        },
        documentoProdotto(documento),
      ],
    },

    {
      id: "errori-comuni",
      titolo: "Gli errori più frequenti",
      minuti: 3,
      sommario: "Quelli che costano tempo davvero, e come si evitano.",
      blocchi: [
        {
          tipo: "tabella",
          intestazioni: ["Errore", "Conseguenza", "Come si evita"],
          righe: [
            [
              "Lavorare sull'azienda o sull'esercizio sbagliato",
              "Dati corretti nel posto sbagliato: nessun avviso, perché sono validi",
              "Guardare i due selettori in cima a ogni sessione",
            ],
            [
              "Lasciare vuoto invece di scrivere zero",
              "La verifica lo conta come «non rilevato», e il documento lo dichiara mancante",
              "Zero esplicito quando il dato è nullo; vuoto solo se la voce non esiste",
            ],
            [
              "Non compilare l'anno di confronto",
              "Il documento esce senza delta: dice cosa fai, non se stai migliorando",
              "Compilare la colonna del confronto insieme a quella dell'esercizio",
            ],
            [
              "Pubblicare per «vedere come viene»",
              "Una versione in più nell'archivio, che resta lì per sempre",
              "L'anteprima si guarda dal percorso; si pubblica quando è finito",
            ],
          ],
        },
      ],
    },
  ];
}

/**
 * La sezione sull'esercizio, che cambia forma secondo il modulo.
 *
 * ⚠️ Due percorsi su ${NUMERI.moduli} NON sono annuali: l'autovalutazione e la
 * Dichiarazione di Applicabilità sono fotografie che si aggiornano per revisioni. Una
 * sezione unica che parlasse di «esercizio» sarebbe falsa proprio per loro, ed è il tipo di
 * dettaglio che chi legge nota subito — perché sta guardando quella schermata mentre legge.
 */
function esercizio(perEsercizio: boolean): Sezione {
  if (perEsercizio) {
    return {
      id: "esercizio",
      titolo: "L'esercizio, e il confronto con l'anno prima",
      minuti: 3,
      sommario: "Questo percorso si redige per anno, e il confronto è ciò che lo rende leggibile.",
      blocchi: [
        {
          tipo: "prosa",
          testo:
            `Il selettore in cima dice su quale anno stai lavorando. I dati sono per esercizio: cambiando anno ` +
            `cambia tutto quello che vedi sotto.`,
        },
        {
          tipo: "avviso",
          tono: "attenzione",
          titolo: "L'anno di confronto non è facoltativo",
          testo:
            `Senza i dati dell'anno precedente il documento esce con le colonne dei delta vuote: dice quanto ` +
            `consumi, non se stai migliorando. È la differenza fra una fotografia e una misura, e la nota se ne ` +
            `accorge chi legge il documento, non chi lo compila.`,
        },
      ],
    };
  }
  return {
    id: "revisione",
    titolo: "Una fotografia, non un esercizio",
    minuti: 2,
    sommario: "Questo percorso non si redige per anno: si aggiorna, e ogni aggiornamento è una revisione.",
    blocchi: [
      {
        tipo: "prosa",
        testo:
          `A differenza dei percorsi annuali, qui non c'è un esercizio da scegliere: c'è **lo stato corrente**, che ` +
          `si aggiorna quando qualcosa cambia. Pubblicando ottieni una revisione numerata, e le precedenti restano ` +
          `consultabili.`,
      },
      {
        tipo: "avviso",
        tono: "nota",
        titolo: "Quando conviene pubblicare",
        testo:
          `Quando qualcuno deve riceverlo: un committente, un ente di certificazione, una banca. Fra una consegna ` +
          `e l'altra si lavora sullo stato corrente senza pubblicare.`,
      },
    ],
  };
}

/** Che cosa produce questo percorso, detto col nome vero del documento. */
function documentoProdotto(tipo: string | undefined) {
  if (!tipo) {
    return {
      tipo: "avviso" as const,
      tono: "nota" as const,
      titolo: "Questo percorso non produce ancora un documento",
      testo:
        `È l'unico dei ${NUMERI.moduli}: il lavoro che ci fai alimenta gli altri percorsi, e i suoi documenti ` +
        `arrivano nelle fasi conclusive. Detto qui perché una sezione muta, in mezzo ad altre che nominano ` +
        `un'uscita, si legge come una svista.`,
    };
  }
  return {
    tipo: "prosa" as const,
    testo:
      `Il documento che questo percorso produce si trova poi nell'archivio, filtrabile per tipo e per azienda, e ` +
      `si può condividere con il cliente tramite un collegamento che non richiede account.`,
  };
}

File: audio-formazione/distilla-slide/comuni.py
# Distillazione delle slide delle SEZIONI COMUNI (una traccia sola, riusata da tutti i corsi).
# `inizia` NON si scrive a mano: si ricava dal copione, così non può divergere.
#
# ⚠️ Tre regole che qui valgono più che altrove:
# - nessuna chiusura porta `prossimo`: la sezione dopo cambia da corso a corso (esercizio o
#   revisione, e dopo gli errori comuni arriva la prima sezione propria del corso);
# - nessun occhiello numerato: il numero della sezione lo conta il renderer;
# - nessuna cifra: i conteggi del catalogo sono segnaposto di NUMERI.
import json
import sys
from pathlib import Path

sys.stdout.reconfigure(encoding="utf-8")
RADICE = Path(r"C:/Users/user/workingnameBilanciotool/audio-formazione")


def P(h, d):
    return {"h": h, "d": d}


SLIDE = {
    "dove-sei": [
        (0, "apertura", dict(
            titolo="Dove sei, e come ci si muove", sottotitolo="Tre livelli, sempre nello stesso ordine",
            testo="Il portafoglio elenca le aziende, il fascicolo raccoglie i {moduli} percorsi in {gruppi} gruppi, il percorso è dove lavori.",
            agenda=["Portafoglio, fascicolo, percorso", "La barra laterale", "La pagina di apertura dello studio",
                    "Disponibili, non da fare", "Colore e icona"])),
        (1, "confronto", dict(
            titolo="La barra laterale cambia con te",
            a={"h": "Nello studio", "sub": "Le voci generali", "punti": ["Portafoglio", "Documenti", "Agenda", "Compensi"]},
            b={"h": "Dentro un'azienda", "sub": "Il suo nome e i suoi percorsi",
               "punti": ["Dal bilancio alla dichiarazione di applicabilità", "Senza tornare indietro"]})),
        (2, "split", dict(
            titolo="Dove cominciare la giornata",
            lead="La pagina di apertura dello studio raccoglie i numeri del portafoglio, i documenti pubblicati di recente e lo scadenzario.",
            punti=[P("Per urgenza", "che cosa faccio adesso"), P("Per cliente", "chi è più indietro"),
                   P("Per ambito", "dove sono indietro")])),
        (3, "evidenza", dict(
            eyebrow="Per chi apre il fascicolo la prima volta",
            titolo="{moduli} percorsi disponibili, non {moduli} da fare.",
            testo="Si aprono quelli che servono a quel cliente. Gli altri restano spenti, e non compaiono nello scadenzario.")),
        (4, "definizione", dict(
            eyebrow="Non è decorazione", titolo="Colore e icona",
            definizione="Il modo più rapido per capire dove sei quando hai otto clienti aperti e le schermate cominciano a somigliarsi.",
            punti=[P("Il colore", "dice di che materia si parla"), P("L'icona", "dice quale lavoro stai facendo")])),
    ],
    "come-si-salva": [
        (0, "apertura", dict(
            titolo="Come si salva", sottotitolo="Da solo, campo per campo",
            testo="Il pulsante salva non c'è: ogni campo si salva nel momento in cui lo lasci, e il dato è sul server.",
            agenda=["Il salvataggio automatico", "Per campo, non per riga", "Dati isolati fra studi", "Niente da esportare"])),
        (1, "split", dict(
            titolo="Per campo, non per riga",
            lead="Una scelta pagata con quattro difetti veri: la pagina rimandava tutta la riga, compresa la parte non ancora aggiornata.",
            punti=[P("Salvare il costo", "azzerava la quantità appena scritta"),
                   P("La rilevanza finanziaria", "cancellava l'impatto messo un attimo prima"),
                   P("Adesso torna un campo solo", "e il valore precedente lo rilegge il server")])),
        (2, "cards", dict(
            titolo="Due cose sui dati",
            cards=[P("Isolati fra studi", "Il divieto lo applica il database: un altro studio non li legge nemmeno per un nostro errore di programmazione."),
                   P("Condivisi dentro lo studio", "I colleghi che inviti vedono tutto il portafoglio, così un socio può sostituire l'altro.")])),
        (3, "evidenza", dict(
            eyebrow="Il cambiamento più grande",
            titolo="Nessun archivio da esportare, nessun file da conservare.",
            testo="Il lavoro sta sul server: cambiare computer significa solo rifare l'accesso.")),
    ],
    "esercizio": [
        (0, "apertura", dict(
            titolo="L'esercizio, e il confronto con l'anno prima", sottotitolo="Questo percorso si redige per anno",
            testo="Il selettore in cima dice su quale esercizio lavori: i dati sono per anno, e l'anno vecchio resta com'era.",
            agenda=["Il selettore dell'esercizio", "Copiare la struttura dell'anno prima", "Il confronto non è facoltativo",
                    "Le due colonne insieme", "Il primo anno"])),
        (1, "evidenza", dict(
            eyebrow="Da ricordare",
            titolo="Senza l'anno prima, il documento dice quanto consumi, non se migliori.",
            testo="Esce lo stesso, con le colonne delle variazioni vuote. E chi lo legge cerca proprio quella colonna.")),
        (2, "split", dict(
            titolo="Le due colonne, insieme",
            lead="Compilarle nello stesso momento costa molto meno che tornarci fra un mese.",
            punti=[P("Le fatture", "sono già aperte"), P("Il referente", "è già al telefono"), P("La visura", "è già sul tavolo")])),
        (3, "definizione", dict(
            eyebrow="Il primo anno", titolo="Il confronto che non c'è, dichiarato",
            definizione="Se è la prima volta che segui l'azienda, il documento dichiara che manca l'anno di confronto invece di lasciare un buco.",
            punti=[P("Dall'anno dopo", "il confronto c'è per costruzione")])),
    ],
    "revisione": [
        (0, "apertura", dict(
            titolo="Una fotografia, non un esercizio", sottotitolo="Si aggiorna, e si consegna per revisioni",
            testo="Niente selettore dell'anno: c'è lo stato corrente, e il lavoro prosegue invece di ricominciare a gennaio.",
            agenda=["Lo stato corrente", "Quando pubblicare", "Aggiornare non è consegnare"])),
        (1, "cards", dict(
            titolo="Si pubblica quando qualcuno deve riceverlo",
            cards=[P("Un committente", "che lo ha chiesto"), P("Un ente di certificazione", "o una banca"),
                   P("Un cliente a monte", "della filiera")])),
        (2, "confronto", dict(
            titolo="Aggiornare non è consegnare",
            a={"h": "Aggiornare", "sub": "Continuo", "punti": ["Costa poco per volta", "Si lavora sullo stato corrente"]},
            b={"h": "Consegnare", "sub": "Un momento scelto",
               "punti": ["Produce una revisione numerata", "Le precedenti restano consultabili"]})),
    ],
    "la-verifica": [
        (0, "apertura", dict(
            titolo="La verifica prima di consegnare", sottotitolo="L'elenco di ciò che manca",
            testo="Non è un riassunto e non è un voto: ogni voce mancante è un collegamento al punto da sistemare.",
            agenda=["Che cosa elenca", "Completezza, non correttezza", "Che cosa resta a te", "L'ordine giusto"])),
        (1, "definizione", dict(
            eyebrow="La cosa più importante", titolo="Completezza, non correttezza",
            definizione="La verifica guarda se un dato c'è, non se è giusto: un numero sbagliato ma presente risulta a posto, e la riga diventa verde.")),
        (2, "punti", dict(
            titolo="Resta lavoro tuo, anche quando è tutto verde",
            intro="Sbagli di dieci volte il consumo di gas: il documento esce senza nessun avviso.",
            punti=[P("La quadratura", "con le fatture"), P("La plausibilità", "dei fattori di carico"),
                   P("La coerenza", "dei salti da un anno all'altro")])),
        (3, "chiusura", dict(
            titolo="L'ultimo controllo, non il primo",
            punti=[P("Prima i numeri", "e ti chiedi se hanno senso"),
                   P("Poi la verifica", "per non dimenticare una casella"),
                   P("Al contrario", "ti dirà che va tutto bene, e non servirà a niente")])),
    ],
    "pubblicare": [
        (0, "apertura", dict(
            titolo="Che cosa succede quando pubblichi", sottotitolo="Il documento si congela",
            testo="Dati e calcoli in una versione numerata, che il database non lascia più modificare: è il motivo per cui il documento vale.",
            agenda=["Congelare, non stampare", "Quattro cose in quel momento", "Il PDF archiviato", "Non si pubblica per vedere"])),
        (1, "cards", dict(
            titolo="Quattro cose, in quel momento",
            cards=[P("Numero di versione", "Ripubblicando ottieni la seconda, e la prima resta consultabile."),
                   P("Marchio", "Si sceglie lì: i documenti consegnati non cambiano se lo studio cambia logo."),
                   P("Edizione dei contenuti", "Congelata, perché le norme si aggiornano."),
                   P("Codice di verifica", "Chi ha il documento ne controlla l'autenticità dal sito, senza account.")])),
        (2, "evidenza", dict(
            eyebrow="Il PDF",
            titolo="Riscaricarlo restituisce lo stesso identico file.",
            testo="Si genera una volta e resta archiviato: quello consegnato al cliente non cambia impaginazione, nemmeno fra due anni.")),
        (3, "confronto", dict(
            titolo="Anteprima o pubblicazione",
            a={"h": "Anteprima", "sub": "Dal percorso", "punti": ["Quante volte vuoi", "Mentre il lavoro è in corso"]},
            b={"h": "Pubblicazione", "sub": "Quando il lavoro è finito",
               "punti": ["Resta nell'archivio per sempre", "Il cliente vede anche le versioni di prova"]})),
    ],
    "errori-comuni": [
        (0, "apertura", dict(
            titolo="Gli errori più frequenti", sottotitolo="Distrazioni, non incomprensioni",
            testo="Il primo non produce nessun avviso: prima di cominciare, guarda i due selettori dell'azienda e dell'esercizio.",
            agenda=["Azienda o esercizio sbagliato", "Vuoto invece di zero", "L'anno di confronto", "Pubblicare per vedere"])),
        (1, "confronto", dict(
            titolo="Vuoto non è zero",
            a={"h": "Vuoto", "sub": "La voce non esiste", "punti": ["Esce dai conti", "Il documento la dichiara mancante"]},
            b={"h": "Zero", "sub": "Esiste e vale zero", "punti": ["Resta nei conti", "È un'informazione"]})),
        (2, "evidenza", dict(
            eyebrow="L'anno di confronto",
            titolo="Senza confronto è una fotografia, non una misura.",
            testo="Il documento esce con le variazioni vuote: se ne accorge chi lo legge, non chi lo compila.")),
        (3, "tabella", dict(
            titolo="I quattro errori, e come si evitano",
            cols=["Errore", "Conseguenza", "Come si evita"],
            righe=[["Azienda o esercizio sbagliato", "Dati validi nel posto sbagliato, senza avviso", "Guardare i due selettori"],
                   ["Vuoto invece di zero", "Il dato risulta mancante", "Zero quando vale zero"],
                   ["Anno di confronto vuoto", "Nessuna variazione nel documento", "Compilare le due colonne insieme"],
                   ["Pubblicare per vedere", "Una versione in più, per sempre", "Usare l'anteprima"]])),
    ],
}


def main():
    copione = json.loads((RADICE / "_comuni" / "script.json").read_text(encoding="utf-8"))
    paragrafi = {s["id"]: [p.strip() for p in s["script"].split("\n\n") if p.strip()] for s in copione["sezioni"]}
    voci = []
    for sez, elenco in SLIDE.items():
        if sez not in paragrafi:
            raise SystemExit(f"sezione sconosciuta _comuni/{sez}")
        for p, layout, campi in elenco:
            par = paragrafi[sez]
            if p >= len(par):
                raise SystemExit(f"_comuni/{sez}: p {p} oltre i {len(par)} paragrafi")
            inizia = " ".join(par[p].split()[:6])
            voci.append({"sezione": sez, "p": p, "inizia": inizia, "layout": layout, **campi})
    mancanti = set(paragrafi) - set(SLIDE)
    if mancanti:
        raise SystemExit(f"sezioni comuni senza slide: {sorted(mancanti)}")
    (RADICE / "_comuni" / "slide.json").write_text(json.dumps(voci, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    conta = {}
    for v in voci:
        conta[v["layout"]] = conta.get(v["layout"], 0) + 1
    print(f"_comuni: {len(voci)} slide · " + ", ".join(f"{k} {v}" for k, v in sorted(conta.items(), key=lambda x: -x[1])))


main()
